### PR TITLE
Fix PatternMatcher caching in CompletionHelper

### PR DESCRIPTION
**Pull request by genlu**

### CODE REVIEW

File: src/Features/Core/Portable/Completion/PatternMatchHelper.cs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using Microsoft.CodeAnalysis.PatternMatching;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal sealed class PatternMatchHelper : IDisposable
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/PatternMatchHelper.cs:15]: New helper type to handle PatternMatcher related functionality, extracted from CompletionHelper. Not thread safe and must be disposed after use

File: src/Features/Core/Portable/Completion/CompletionHelperServiceFactory.cs [hunk 2]
```diff
@@ -1,81 +0,0 @@
...
-        }
-
-        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
-            => new Service(workspaceServices.Workspace);
-
-        private sealed class Service : ICompletionHelperService, IWorkspaceService
-        {
-            private readonly object _gate = new();
-
-            private CompletionHelper? _lazyCaseSensitiveInstance;
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/CompletionHelperServiceFactory.cs:31]: Singleton now moved into CompletionHelper itself. So this service is no longer needed

File: src/Features/Core/Portable/Completion/Utilities.cs [hunk 3]
```diff
@@ -45,5 +48,14 @@ public static bool IsPreferredItem(this CompletionItem completionItem)
             => completionItem.DisplayText.StartsWith(UnicodeStarAndSpace);
 
         public const string UnicodeStarAndSpace = "\u2605 ";
+
+        public static async Task<SyntaxContext> CreateSyntaxContextWithExistingSpeculativeModelAsync(Document document, int position, CancellationToken cancellationToken)
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Utilities.cs:52]: Code moved here from CompletionHelper

File: src/Features/Core/Portable/Completion/CompletionHelper.cs [hunk 4]
```diff
@@ -3,134 +3,32 @@
...
-using Microsoft.CodeAnalysis.Text;
-using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Completion
 {
     internal sealed class CompletionHelper
     {
-        private readonly object _gate = new();
-        private readonly Dictionary<(string pattern, CultureInfo, bool includeMatchedSpans), PatternMatcher> _patternMatcherMap =
-             new();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs:25]: nice

File: src/Features/Core/Portable/Completion/PatternMatchHelper.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/PatternMatchHelper.cs:15]: doc the threadsafety/dispose part on the type itself :)

File: src/Features/Core/Portable/Completion/CompletionHelper.cs [hunk 5]
```diff
@@ -3,134 +3,32 @@
...
 using Microsoft.CodeAnalysis.Tags;
-using Microsoft.CodeAnalysis.Text;
-using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Completion
 {
     internal sealed class CompletionHelper
     {
-        private readonly object _gate = new();
-        private readonly Dictionary<(string pattern, CultureInfo, bool includeMatchedSpans), PatternMatcher> _patternMatcherMap =
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs:24]: We were caching PatternMatchers for all the patterns given to us, which means at least one instance for each keystroke...

**Pull request merged**